src/mono/wasm/test-main.js

```diff
@@ -143,7 +143,7 @@ loadDotnet("./dotnet.js").then((createDotnetRuntime) => {
     return createDotnetRuntime(({ MONO, INTERNAL, BINDING, Module }) => ({
         disableDotnet6Compatibility: true,
         config: null,
-        configSrc: "./mono-config.json",
+        configSrc: processedArguments.run_deep_working_dir ? "./AppBundle/mono-config.json" : "./mono-config.json",
```

Comment:
Perhaps 
```
await mono_wasm_load_config(locateFile(moduleExt.configSrc));
```
In  `mono_wasm_pre_init` of startup.ts would be better ?

Reply:
That sounds like a good change.
However, if we adopt that modification, I think it will be difficult to determine from where the relative path in the jsdoc comment in `configFilePath` is relative.

https://github.com/dotnet/runtime/blob/97b29919e4bb7eaebf4c5cdb6b53bce3b459a0c3/src/mono/wasm/runtime/startup.ts#L668

What do you think of the idea of adding a statement that the path is relative to dotnet.js?

Reply:
@radical I know you are thinking about redesign of the whole area. Could you please add your thoughts ?

Reply:
I'm not looking for quick hack,  but rather approach, which would be easy to understand and which would be consistent on all targets.

Reply:
I'm thinking about naïve approach from user perspective:
- `configSrc` is relative to `dotnet.js` unless I use absolute path
- `config.assembly_root` is relative to `configSrc` unless I use absolute path
- `asset.name` is relative to `configSrc` unless I use absolute path
- `sourcePrefix` is used only for relative paths.

Does this make sense ? Or there is better alternative ?
I admit that I didn't explore all consequences now. Is any of that breaking change ?

Why `scriptDirectory` didn't work on V8 ? 

I know we had trouble resolving it on ES6. See
https://github.com/dotnet/runtime/blob/bfbb78354e536ac616f2f9dabf3db2b8fa8b9f64/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L24
It only happen asynchronously. But that should be fine.

New approach has to work for ES6 too. 

And ideally for webpack too.
See https://github.com/webpack/webpack/issues/7352
And https://github.com/webpack/webpack/pull/15246

Reply:
On `V8` even if the `scriptDirectory` is always empty, could we treat it as `.` and do relative paths with it ? 
This worked for me d8> `read("../xx.md")`. 

Which is what our `fetch_like` does.  See https://github.com/dotnet/runtime/blob/973b7640aed5f6f1aa95bea10361b7792f9c1c5d/src/mono/wasm/runtime/polyfills.ts#L32

Reply:
That makes sense.

Here is an example of breaking changes in Node.js environment.

Current main

- `configSrc` is relative to `cwd` unless I use absolute path
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name` (assembly and resource) is relative to `dotnet.js` unless I use absolute path
  - `asset.name` (other) is relative to `cwd` unless I use absolute path
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

Patch https://github.com/dotnet/runtime/pull/69441/commits/a1c3c7a3651f14266b9f4bf48e9d81775303de7b

- `configSrc` is relative to `cwd` unless I use absolute path
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name` (assembly and resource) is relative to `dotnet.js` unless I use absolute path
  - `asset.name` (other) is relative to **`dotnet.js`** unless I use absolute path <- changed, this seems reasonable to me.
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

And https://github.com/dotnet/runtime/pull/69441/commits/838852557c4988b5dcae2e639390c3ef5e3dc439

- `configSrc` is relative to **`dotnet.js`** unless I use absolute path <- changed
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name`  is relative to `dotnet.js` unless I use absolute path
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

Thus, the path to some `cwd` is changed to the path to `dotnet.js`.

On V8, everything relative to `cwd` except `sourcePrefix`.

The reason can be led from several things.
When executed with the following command, the variables indicating the environment are as follows.

```sh
$ v8 test-main.js
```

- ENVIRONMENT_IS_WEB == false
- ENVIRONMENT_IS_WORKER == false
- ENVIRONMENT_IS_NODE == false
- ENVIRONMENT_IS_SHELL == true

Looking at the emscripten code, there is no value set for `scriptDirectory` in the case of `ENVIRONMENT_IS_SHELL`.

https://github.com/emscripten-core/emscripten/blob/e1d97f55d12ed63dc30e547c818af9a63be0b976/src/shell.js#L155

So the behavior of `locateFile` is simply to return the filename, which is resolved by the path to the cwd.
Even if `./` is given at the beginning of the file and treated as a relative path, it will be the path to the cwd.

However, when `dotnet.js` is loaded with `import` statement, `import.meta.url` can be retrieved, so I am thinking that it may be possible to override the `scriptDirectory` by combining it with the following code. (like https://github.com/dotnet/runtime/blob/bfbb78354e536ac616f2f9dabf3db2b8fa8b9f64/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L24)

```js
if (ENVIRONMENT_IS_SHELL)  {
  const tmp = import.meta.url;
  scriptDirectory = tmp.substring(0, tmp.lastIndexOf('/') + 1) + '/';
}
```

The combination of CommonJS and V8 would be the path to cwd unless an external base path is given...

By the way, I have not been able to research the web (Chrome).
This is because the following function fails to load `dotnet.js` and `wasm`.

https://github.com/dotnet/runtime/blob/a1c3c7a3651f14266b9f4bf48e9d81775303de7b/src/mono/wasm/runtime/startup.ts#L28

This is because the code below specifies the `href` as `dotnet.js` and uses `location.href `as the starting point unless the `base` tag is used.
https://github.com/dotnet/runtime/blob/a1c3c7a3651f14266b9f4bf48e9d81775303de7b/src/mono/wasm/runtime/startup.ts#L46

Also, the behavior through the bundler (like webpack, rollup...) has not been investigated.

Reply:
I have investigated whether this change would work under ES6 environment.

In conclusion, it failed regardless of this PR change.
We need to change the Assertion in the following places to test under ES6 environment.

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs#L539
(If WasmEnableES6 is true, re-link will occur and the Assertion at that point will fail according to the Config.)

The following changes would also be required.

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/mono/wasm/test-main.js#L259-L261
(If test-main.js is executed as type: module, dotnet.js needs to be loaded with `import` statement.)

https://github.com/emscripten-core/emscripten/blob/af7f969c4497e642c4f7c256531f93b89a8b008a/src/preamble.js#L786
(URL is not defined in V8 Shell. It's difficult to bypass...)

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L50-L51
(If `IS_SHELL`, `scriptDirectory` must be overwritten using `import.meta.url`)

Thus, we have a lot of interest in the changes we want to make.
If we are also going to test in an ES6 environment, I thought we needed to resolve https://github.com/dotnet/runtime/issues/63337

Reply:
Since https://github.com/dotnet/runtime/pull/70746  was merged, I would consider testing again.

Reply:
https://github.com/dotnet/runtime/pull/69441#discussion_r882945695

I wrote as "configSrc is relative to dotnet.js unless I use absolute path <- changed" in this reply, but it was different.
If configSrc is given as an absolute path, the current change does not resolve the configSrc path correctly...
https://github.com/dotnet/runtime/pull/69441/commits/838852557c4988b5dcae2e639390c3ef5e3dc439

https://github.com/dotnet/runtime/pull/69441/files#diff-3e3e5eb583ed48d37f0ab14c4060488a683d4088eff83a58be530cfd43fcb213R31-R54

For this change, if path is an absolute path or uri containing a protocol such as https, we need to make a change to return only path.

like this
```js
const toAbsoluteUrl = function toAbsoluteUrl(path, prefix) {
  ...
  if (isAbsoluteOrUriOr....(path)) {
    return path;
  }
}
```

Or, if it is determined that the path is not relative, it is necessary to treat it as an absolute path and fetch it instead of using `locateFile`.

https://github.com/dotnet/runtime/pull/69441/files#diff-5cdf20fb11670158ec01622e5d8e6764608b9846dfd31ac45c85bde7f0ed4228R687

Reply:
https://github.com/dotnet/runtime/pull/69441#discussion_r882394336

As pointed out there, it seems to me that `config.assembly_root` , etc. should be relative to `configSrc`.
My current PR does not support this behavior because it is relative to `dotnet.js`.

As a hack, this can be achieved by replacing `Module["locateFile"]` with a function relative to `configSrc` in `onConfigLoaded`.
I have also tried wrapping the `locateFile` used in `mono_download_assets` and using `configSrc` as the origin, but I cannot determine the best way to do this...

Reply:
In my PR it's relative to `dotnet.js`. Maybe that's good enough ? People could still put absolute path there.

Reply:
I think we can start with a relative path in `dotnet.js` .
I am assuming this is for web and am wondering what happens if I use a bundler (e.g. webpack or vite).
But I think it is good to leave that out of scope at this stage.

Reply:
There is range of issues with webpack+dotnet (emscripten). One of them is dynamic `import("something")` and another `import.meta.url`.

There is my last attempt to make it work 
https://github.com/dotnet/runtime/blob/main/src/mono/sample/wasm/browser-webpack/webpack.config.js#L9

Reply:
Thanks for the suggestions on how to deal with this.
I thought dotnet.js would be a good starting point since the means seems to exist.